File: frontend/src/features/codex-reset-forecast/schemas.ts
import { z } from "zod";

export const ResetSignalKindSchema = z.enum([
  "explicit_reset_announcement",
  "tibo_ok_request",
  "sam_trigger",
  "issue_acknowledged",
  "limit_fix",
  "confirmed_reset",
  "no_active_signal",
]);

export const CodexResetEvidenceSchema = z.object({
  kind: ResetSignalKindSchema,
  label: z.string(),
  summary: z.string(),
  observedAt: z.string().nullable(),
  url: z.string().nullable(),
  source: z.string(),
  ageHours: z.number().nullable(),
  score: z.number(),
});

export const CodexResetScoreFactorSchema = z.object({
  key: z.string(),
  label: z.string(),
  value: z.number(),
  description: z.string(),
});

export const CodexResetHistoricalExampleSchema = z.object({
  label: z.string(),
  classification: z.string(),
  signalAt: z.string(),
  resetAt: z.string(),
  leadTimeHours: z.number(),
  signalSummary: z.string(),
  resetSummary: z.string(),
  signalUrl: z.string(),
  resetUrl: z.string(),
});

export const CodexResetCollectionStatusSchema = z.object({
  refreshEnabled: z.boolean(),
  refreshInProgress: z.boolean(),
  lastStartedAt: z.string().nullable(),
  lastCompletedAt: z.string().nullable(),
  lastError: z.string().nullable(),
  nextRefreshDueAt: z.string().nullable(),
});

export const CodexResetXActivityItemSchema = z.object({
  authorHandle: z.string(),
  kind: z.enum(["post", "reply", "quote"]),
  observedAt: z.string(),
  text: z.string(),
  translatedTextZh: z.string(),
  url: z.string(),
  replyTo: z.string().nullable(),
  parent: z
    .object({
      authorHandle: z.string(),
      text: z.string(),
      translatedTextZh: z.string(),
      url: z.string().nullable(),
    })
    .nullable(),
  resetRelevance: z.enum(["none", "weak", "strong"]),
  relevanceSummary: z.string(),
});

export const CodexResetForecastSchema = z.object({
  generatedAt: z.string(),
  horizonHours: z.number(),
  probability: z.number(),
  probabilityPercent: z.number(),
  probabilityLevel: z.enum(["low", "elevated", "high"]),
  confidence: z.enum(["low", "medium", "high"]),
  summary: z.string(),
  modelVersion: z.string(),
  sourceNote: z.string(),
  collectionStatus: CodexResetCollectionStatusSchema,
  currentEvidence: z.array(CodexResetEvidenceSchema),
  latestXItems: z.array(CodexResetXActivityItemSchema),
  scoreFactors: z.array(CodexResetScoreFactorSchema),
  historicalExamples: z.array(CodexResetHistoricalExampleSchema),
});

export type CodexResetEvidence = z.infer<typeof CodexResetEvidenceSchema>;
export type CodexResetHistoricalExample = z.infer<typeof CodexResetHistoricalExampleSchema>;
export type CodexResetXActivityItem = z.infer<typeof CodexResetXActivityItemSchema>;
export type CodexResetForecast = z.infer<typeof CodexResetForecastSchema>;

File: app/modules/codex_reset_forecast/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.modules.shared.schemas import DashboardModel

ResetSignalKind = Literal[
    "explicit_reset_announcement",
    "tibo_ok_request",
    "sam_trigger",
    "issue_acknowledged",
    "limit_fix",
    "confirmed_reset",
    "no_active_signal",
]
ProbabilityLevel = Literal["low", "elevated", "high"]
ForecastConfidence = Literal["low", "medium", "high"]
CodexResetXItemKind = Literal["post", "reply", "quote"]
CodexResetXItemRelevance = Literal["none", "weak", "strong"]


class CodexResetEvidence(DashboardModel):
    kind: ResetSignalKind
    label: str
    summary: str
    observed_at: datetime | None = None
    url: str | None = None
    source: str
    age_hours: float | None = Field(default=None, ge=0.0)
    score: float = Field(ge=0.0, le=1.0)


class CodexResetScoreFactor(DashboardModel):
    key: str
    label: str
    value: float = Field(ge=0.0, le=1.0)
    description: str


class CodexResetHistoricalExample(DashboardModel):
    label: str
    classification: str
    signal_at: datetime
    reset_at: datetime
    lead_time_hours: float = Field(gt=0.0)
    signal_summary: str
    reset_summary: str
    signal_url: str
    reset_url: str


class CodexResetCollectionStatus(DashboardModel):
    refresh_enabled: bool
    refresh_in_progress: bool
    last_started_at: datetime | None = None
    last_completed_at: datetime | None = None
    last_error: str | None = None
    next_refresh_due_at: datetime | None = None


class CodexResetXActivityItem(DashboardModel):
    class ParentItem(DashboardModel):
        author_handle: str
        text: str
        translated_text_zh: str
        url: str | None = None

    author_handle: str
    kind: CodexResetXItemKind
    observed_at: datetime
    text: str
    translated_text_zh: str
    url: str
    reply_to: str | None = None
    parent: ParentItem | None = None
    reset_relevance: CodexResetXItemRelevance
    relevance_summary: str


class CodexResetForecastResponse(DashboardModel):
    generated_at: datetime
    horizon_hours: int = Field(gt=0)
    probability: float = Field(ge=0.0, le=1.0)
    probability_percent: int = Field(ge=0, le=100)
    probability_level: ProbabilityLevel
    confidence: ForecastConfidence
    summary: str
    model_version: str
    source_note: str
    collection_status: CodexResetCollectionStatus
    current_evidence: list[CodexResetEvidence]
    latest_x_items: list[CodexResetXActivityItem]
    score_factors: list[CodexResetScoreFactor]
    historical_examples: list[CodexResetHistoricalExample]
